File: character-sheet-companion-server/src/listeners/actor/actorBaseDataListener.ts
import { Server, Socket } from "socket.io";
import {
  getFoundrySocketFromRoom,
  getIosSocketFromRoom,
} from "../../utilities/SocketUtilities";
import { SOCKET_EVENTS } from "../../constants/events";
import { ActorBaseData } from "../../interfaces/actor/ActorBaseData";

export const actorBaseDataListener = (io: Server, socket: Socket): void => {
  const requestActorBaseData = (actorId: string) => {
    const foundrySocket = getFoundrySocketFromRoom(io, socket);
    if (foundrySocket) {
      foundrySocket.emit(
        SOCKET_EVENTS.SERVER.ACTOR.REQUEST.REQUEST_ACTOR_BASE_DATA,
        actorId,
        socket.id
      );
    }
  };

  const receiveActorBaseData = (
    actorBaseData: ActorBaseData,
    iosSocketId: string
  ) => {
    const iosSocket = getIosSocketFromRoom(io, socket, iosSocketId);
    if (iosSocket) {
      iosSocket.emit(
        SOCKET_EVENTS.SERVER.ACTOR.SEND.SEND_ACTOR_BASE_DATA,
        JSON.stringify(actorBaseData)
      );
    }
  };

  socket.on(
    SOCKET_EVENTS.IOS.ACTOR.REQUEST_ACTOR_BASE_DATA,
    requestActorBaseData
  );
  socket.on(
    SOCKET_EVENTS.FOUNDRY.ACTOR.SEND_ACTOR_BASE_DATA,
    receiveActorBaseData
  );
};

File: CharacterSheetCompanion/CharacterSheetCompanion/Services/SocketListeners/ActorListeners/ActorBaseDataListener.swift
//
//  ActorBaseDataListener.swift
//  CharacterSheetCompanion
//
//  Created by Zachary Sanders on 8/17/21.
//

import Combine
import Foundation
import SocketIO

class ActorBaseDataListener: SocketListener, ActorListener {
    let baseDataPublisher: AnyPublisher<ActorBaseDataModel?, Never>
    let socket: SocketIOClient

    private let baseDataSubject = CurrentValueSubject<ActorBaseDataModel?, Never>(nil)

    init(socket: SocketIOClient) {
        baseDataPublisher = baseDataSubject.eraseToAnyPublisher()
        self.socket = socket
    }

    func addSocketHandlers() {
        socket.on(SocketEvents.SERVER.ACTOR.SEND.SEND_ACTOR_BASE_DATA) { data, _ in
            do {
                try self.baseDataSubject.send(SocketListenerUtility.parseSocketEventData(data))
            } catch let FoundryJSONError.errorMessage(errorMessage) {
                print(errorMessage)
            } catch {
                print(error)
            }
        }
    }

    func requestInitialActorData(actorId: String) {
        socket.emit(SocketEvents.IOS.ACTOR.REQUEST_ACTOR_BASE_DATA, actorId)
    }

    func isReady() -> Bool {
        true
    }
}

extension SocketEvents.IOS.ACTOR {
    static let REQUEST_ACTOR_BASE_DATA = "ios:requestActorBaseData"
}

extension SocketEvents.SERVER.ACTOR.SEND {
    static let SEND_ACTOR_BASE_DATA = "server:sendActorBaseData"
}
